### PR TITLE
Support empty tensor input for some models.

### LINKED ISSUE #2193: dist_train keep waiting when filter_empty_gt=False
Environment:
* pytorch 1.3.1
* cuda 10.0
* cudnn 7
* mmdetection 1.1.0

My config file:  cascade_rcnn_dconv_c3-c5_r50_fpn_1x.py,  **with only the ann path modified**.

1. When i set `filter_empty_gt=False` (20% of my images are background only.),  there are not any error messages, but the trainning process are always waiting..., so I have to keyboardinterupt it: 

image

2. When i set `filter_empty_gt=True`, every thing is ok.

### PR DESCRIPTION
**Pull request by yhcao6**
Fix #2193

### CODE REVIEW

File: mmdet/ops/wrappers.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+class ConvTranspose2d(nn.ConvTranspose2d):
+
+    def forward(self, x):
+        if x.numel() == 0 and TORCH_VERSION <= (1, 4):
+            out_shape = list(x.shape[:2])
+            for i, k, p, s, d, op in zip(x.shape[-2:], self.kernel_size,
+                                         self.padding, self.stride,
+                                         self.dilation, self.output_padding):
+                out_shape.append((i - 1) * s - 2 * p + (d * (k - 1) + 1) +
+                                 2 * op)
```
Comment (xvjiarui) [inline mmdet/ops/wrappers.py]: According to PyTorch doc, it should be `1 * op` instead of `2 * op`.

File: tests/test_wrappers.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    test_cases = {
+        'in_w': [10, 20],
+        'in_h': [10, 20],
+        'in_channel': [1, 3],
+        'out_channel': [1, 3],
+        'kernel_size': [3, 5],
+        'stride': [1, 2],
+        'padding': [0, 1],
+        'dilation': [1, 2]
+    }
```
Comment (hellock) [inline tests/test_wrappers.py:22]: Use a list of dict to simplify the following steps.

File: tests/test_wrappers.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+    for in_h, in_w, in_cha, out_cha, k, s, p, d in product(
+            *list(test_cases.values())):
+        # wrapper op with 0-dim input
+        x = torch.empty(0, in_cha, in_h, in_w)
+        wrapper = MaxPool2d(k, stride=s, padding=p, dilation=d)
+        wrapper_out = wrapper(x)
```
Comment (hellock) [inline tests/test_wrappers.py:123]: BP is not tested.

File: mmdet/models/roi_extractors/single_level.py [hunk 4]
```diff
@@ -88,14 +88,17 @@ def roi_rescale(self, rois, scale_factor):
...
         out_size = self.roi_layers[0].out_size
         num_levels = len(feats)
-        target_lvls = self.map_roi_levels(rois, num_levels)
         roi_feats = feats[0].new_zeros(
             rois.size(0), self.out_channels, *out_size)
+
+        if num_levels == 1:
+            if len(rois) == 0:
+                return roi_feats
+            return self.roi_layers[0](feats[0], rois)
```
Comment (hellock) [inline mmdet/models/roi_extractors/single_level.py:99]: What's the intention of checking num_levels?

File: mmdet/core/mask/mask_target.py [hunk 5]
```diff
@@ -8,8 +8,8 @@ def mask_target(pos_proposals_list, pos_assigned_gt_inds_list, gt_masks_list,
     cfg_list = [cfg for _ in range(len(pos_proposals_list))]
     mask_targets = map(mask_target_single, pos_proposals_list,
                        pos_assigned_gt_inds_list, gt_masks_list, cfg_list)
-    mask_targets = torch.cat(list(mask_targets))
-    return mask_targets
+    mask_targets = list(mask_targets)
+    return torch.cat(mask_targets) if len(mask_targets) > 0 else mask_targets
```
Comment (hellock) [inline mmdet/core/mask/mask_target.py:12]: Use a simpler form to ease the code reading:
```python
if len(mask_targets) > 0:
    mask_targets = torch.cat(mask_targets)
return mask_targets
```

File: tests/test_wrappers.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+        ref_out = ref(x)
+
+        assert wrapper_out.shape[0] == 0
+        assert wrapper_out.shape[1:] == ref_out.shape[1:]
+
+        wrapper_out.sum().backward()
+        assert wrapper.weight.grad is not None
+        assert wrapper.weight.grad.shape == wrapper.weight.shape
+
+        x = torch.randn(3, in_cha, in_h, in_w)
```
Comment (hellock) [inline tests/test_wrappers.py:107]: No need to forward twice.

File: tests/test_wrappers.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+        'stride': [1, 2],
+        'padding': [0, 1],
+        'dilation': [1, 2]
+    }
+
+    # train mode
+    for in_h, in_w, in_cha, out_cha, k, s, p, d in product(
+            *list(test_cases.values())):
+        # wrapper op with 0-dim input
+        x = torch.randn(0, in_cha, in_h, in_w)
```
Comment (hellock) [inline tests/test_wrappers.py:28]: `x_empty`

File: tests/test_wrappers.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+        torch.manual_seed(0)
+        wrapper = Conv2d(in_cha, out_cha, k, stride=s, padding=p, dilation=d)
+        wrapper_out = wrapper(x)
```
Comment (hellock) [inline tests/test_wrappers.py:31]: `wrapper_out_empty`

File: tests/test_wrappers.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+
+        # torch op with 3-dim input as shape reference
+        x = torch.randn(3, in_cha, in_h, in_w)
```
Comment (hellock) [inline tests/test_wrappers.py:34]: `x_normal`
